File: 문제풀이/chap6-21_25/prob6-22.py
# 입력
N = int(input())
T = list(map(int, input().split()))

# 배열 초기화
sumT = sum(T)
dp = [ [ False ] * (sumT + 1) for i in range(N + 1) ]
dp[0][0] = True

# 동적계획법
for i in range(1, N + 1):
	for j in range(sumT + 1):
		if j < T[i - 1]:
			if dp[i - 1][j] == True:
				dp[i][j] = True
			else:
				dp[i][j] = False
		if j >= T[i - 1]:
			if dp[i - 1][j] == True or dp[i - 1][j - T[i - 1]] == True:
				dp[i][j] = True
			else:
				dp[i][j] = False

# 답을 계산하고 출력
answer = 10 ** 10
for i in range(sumT + 1):
	if dp[N][i] == True:
		cooking_time = max(i, sumT - i)
		answer = min(answer, cooking_time)
print(answer)

File: 문제풀이/chap6-21_25/prob6-22.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int N, T[109]; bool dp[109][100009];

int main() {
	// 입력
	cin >> N;
	for (int i = 1; i <= N; i++) cin >> T[i];
	
	// 배열 초기화
	int sumT = 0;
	for (int i = 1; i <= N; i++) sumT += T[i];
	for (int i = 1; i <= sumT; i++) dp[0][i] = false;
	dp[0][0] = true;
	
	// 동적계획법
	for (int i = 1; i <= N; i++) {
		for (int j = 0; j <= sumT; j++) {
			if (j < T[i]) {
				if (dp[i - 1][j] == true) dp[i][j] = true;
				else dp[i][j] = false;
			}
			if (j >= T[i]) {
				if (dp[i - 1][j] == true || dp[i - 1][j - T[i]] == true) dp[i][j] = true;
				else dp[i][j] = false;
			}
		}
	}
	
	// 답을 계산하고 출력
	int answer = (1 << 30);
	for (int i = 0; i <= sumT; i++) {
		if (dp[N][i] == true) {
			int cooking_time = max(i, sumT - i);
			answer = min(answer, cooking_time);
		}
	}
	cout << answer << endl;
	
	return 0;
}